README.md
<!--
This README describes the package. If you publish this package to pub.dev,
this README's contents appear on the landing page for your package.

For information about how to write a good package README, see the guide for
[writing package pages](https://dart.dev/guides/libraries/writing-package-pages).

For general information about developing packages, see the Dart guide for
[creating packages](https://dart.dev/guides/libraries/create-library-packages)
and the Flutter guide for
[developing packages and plugins](https://flutter.dev/developing-packages).
-->

This library aims to help developer get started quickly and create a form made out of json.


## Features

The JsonForm can be updated both from the user interaction but can also be notified from outside
(check the example for the stream). You can also get the Json back.

## Usage
You can declare a Json as:
```dart
  final Map<String, dynamic> json = {
    "widgets": [
      {
        "id": "1",
        "name": "Toggle",
        "type": "toggle",
        "values": ["On", "Off"],
        "default_value": "0",
        "chosen_value": 1
      },
      {
        "id": "2",
        "name": "Static text",
        "type": "static_text",
        "chosen_value": "value",
        "description": "(description..)",
      },
      {
        "id": "3",
        "name": "Edit text",
        "type": "edit_text",
        "chosen_value": "edit value",
        "description": "(edit description..)",
      },
      {"type": "header", "name": "Header", "id": "99"},
      {
        "id": "4",
        "name": "Drop down",
        "type": "drop_down",
        "values": ["Low-Intermediate", "Medium", "High"],
        "chosen_value": "Low-Intermediate"
      }
    ]
  };
```

And easily use the basic Theme (or edit it)

```dart
 @override
 Widget build(BuildContext context) {
  return Sizer(
    builder: (BuildContext context, Orientation orientation, DeviceType deviceType) {
      return Scaffold(
        appBar: AppBar(
          title: const Text('Floating Action Button'),
        ),
        body: RefreshIndicator(
          onRefresh: _refresh,
          child: JsonFormWithThemeBuilder(
              jsonWidgets: json)
              .setDateBuilderMethod(dateBuilder)
              .registerComponent(DropDownParser2Creator())
              .setStreamUpdates(onValueChangeStream)
              .setOnValueChanged((String d, dynamic s) async {
            print("Update id $d to value $s");
            await Future.delayed(const Duration(seconds: 1));
            return Future.value(true);
          })
              .setTheme(const DefaultTheme()).build(),
        ),
        floatingActionButton: FloatingActionButton(
          onPressed: () {
            counter++;
            if (counter % 4 == 1) {
              toggle++;
              _onUserController.add({}..["1"] = toggleList[toggle % 2]); // toggle
            }
            if (counter % 4 == 2) {
              _onUserController.add({}..["2"] = "updated" + Random().nextInt(10).toString()); // toggle
            }
            if (counter % 4 == 3) {
              _onUserController.add({}..["3"] = "Val" + Random().nextInt(100000).toString()); // toggle
            }
            if (counter % 4 == 0) {
              _onUserController.add({}..["4"] = list[toggle % 2]); // toggle
            }
          },
          child: const Icon(Icons.navigation),
          backgroundColor: Colors.green,
        ),
      );
    },
  );
}
```
<p align="center">
  <img src="https://raw.githubusercontent.com/itaishalom/JsonToForm/main/screen_shot.jpg" width="350">
</p>


## Additional information

contribute to the package, how to file issues, what response they can expect
from the package authors, and more.
# Created For Aidoc Medical and being used by customers around the world.


CHANGELOG.md
## 0.0.1
Initial release

## 0.0.11
Fix drop down conversion dynamic to string

## 0.0.12
Fix toggle conversion dynamic to string

## 0.0.13
Added debounce_time as an option to edit text

## 0.0.14
Edit text modifications

## 1.0.0
Changed constructor and optional params.
Added tests.
Added support for dynamic widgets + added example of usage.

## 1.0.1
More tests

## 1.0.2
More tests and bug fixes

## 1.0.3
Change color on focus of edit text

## 1.1.0
Add date under the headline!

## 1.1.1
Edit text fix

## 1.1.2
Set value for parser

## 1.2.0
Toggle will now receive the value and not the index

## 1.2.1
Edit text fix

## 1.2.2
Better support long headers,
Edit text is now can be read only,
New themes

## 1.2.3
Fix color of the edit text when unfocus.
New component! Long edit text.

## 1.2.4
Long text will shrink and extend upon the text's length,
Fix description space.

## 1.2.5
Support refresh

## 1.2.6
Support refresh - fix

## 1.2.7
Support refresh - fix2

## 1.2.9
Edit text bug fix

## 1.2.10
Edit text bug fix

## 1.2.11
Time will be updated after conformation
Better log on the exceptions.

## 1.2.12
Support refresh
Better performance

## 1.2.13
Prevent memory leak

## 1.2.14
Header margins added

## 1.2.15
Time will be updated even if started with null;

## 1.2.16
Drop down parser better ui

## 1.2.17
Support keyboard types

## 1.2.18
Build fix

## 1.2.19
Toggle fix

## 1.3.0
Overflow feature - still beta

## 1.3.1
Width of dropdown configurable

## 1.3.2
Width of dropdown configurable changed to double

## 1.4.0
Multiple bug fixes

## 1.4.1
Prevent dispose when widget get out of scroll view

## 2.0.0
Updates and bug fixes + new tests

## 2.0.1
Don't cut the text if edit text on focus

## 2.0.2
Drop Down tests

## 2.0.3
Toggle tests

## 2.1.0
Update toggle library.
New must variable - width of header

## 3.0.0
Update build builder to with Id

## 3.0.1
Use edit text without stream

## 3.0.2
Name widget main Axis alignment

## 3.0.3
Use static text without stream

## 3.0.4
Static value width

## 3.0.5
Static value text alignment

## 3.0.6
Toggle switch update - added activeToggleBorder.

## 4.0.0
Register new widgets to parse with registerComponent.
Thanks for @ItzikAlkotzerAidoc for the PR

## 4.0.1
Fix minimum heights.

## 4.0.2
Fix minimum heights for widgets

## 4.1.0
Remove sizer and bug fix

## 4.1.1
Edit mode - save bar

## 4.1.2
Edit mode - save bar.
Can select between save bar and save with debounce timer

## 4.1.3
Get scrollPadding configuration from theme 

## 4.1.4
Add Padding to scroll when keyboard is open

## 4.1.5
Number of digits before adding ".." is now configurable

## 4.1.6
Number of digits before adding ".." is now configurable on Text with save button

## 4.1.7
Update dependencies

## 4.1.8
Migrate to liveData

## 4.1.9
Update dependencies

## 4.1.10
Update dependencies

example/lib/main.dart
import 'dart:async';
import 'dart:math';
import 'package:example/drop_down_parser2.dart';
import 'package:flutter/material.dart';
import 'package:intl/intl.dart';
import 'package:json_to_form_with_theme/json_to_form_with_theme.dart';
import 'package:json_to_form_with_theme/themes/json_form_theme.dart';
import 'package:sizer/sizer.dart';

void main() {
  runApp(const MyApp());
}

class MyApp extends StatelessWidget {
  const MyApp({Key? key}) : super(key: key);

  // This widget is the root of your application.
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Flutter Demo',
      theme: ThemeData(
        // This is the theme of your application.
        //
        // Try running your application with "flutter run". You'll see the
        // application has a blue toolbar. Then, without quitting the app, try
        // changing the primarySwatch below to Colors.green and then invoke
        // "hot reload" (press "r" in the console where you ran "flutter run",
        // or simply save your changes to "hot reload" in a Flutter IDE).
        // Notice that the counter didn't reset back to zero; the application
        // is not restarted.
        primarySwatch: Colors.blue,
      ),
      home: MyHomePage(title: 'Flutter Demo Home Page'),
    );
  }
}

class MyHomePage extends StatefulWidget {
  MyHomePage({Key? key, required this.title}) : super(key: key) {}

  final String title;

  @override
  State<MyHomePage> createState() => _MyHomePageState();
}

class _MyHomePageState extends State<MyHomePage> {
  Stream<Map<String, dynamic>>? onValueChangeStream;
  final StreamController<Map<String, dynamic>> _onUserController = StreamController<Map<String, dynamic>>();

  // Map<String, WidgetParser> dynamics = {};

  final Map<String, dynamic> json = {
    "widgets": [
      {"type": "header", "name": "Header", "id": "29"},
      {
        "id": "1",
        "name": "DVT",
        "type": "toggle",
        "values": ["On", "Off"],
        "chosen_value": null,
        "time": 1630164109066,
      },
      {"type": "header", "name": "Header2", "id": "39"},
      {
        "id": "56",
        "name": "ADVT",
        "type": "toggle",
        "values": ["On", "Off"],
        "chosen_value": "Off",
        "time": 1630164109056,
      },
      {
        "id": "2",
        "name": "Static text",
        "type": "static_text",
        "chosen_value": "value which is long value",
        "description": "(description..)",
        "time": 1640164109066,
      },
      {
        "id": "26",
        "name": "Long name of the line but a short",
        "type": "static_text",
        "chosen_value": "value",
        "description": "(description..)",
        "time": 1640164109066,
      },
      {
        "id": "3",
        "name": "Blood Pressure with a long long long",
        "type": "edit_text",
        "chosen_value": "A",
        "time": 1640260609562,
        "description": "mmHg - with a long description",
      },
      {"type": "header", "name": "Header", "id": "99"},
      {
        "id": "4",
        "name": "Drop down",
        "type": "drop_down",
        "time": 1640264109066,
        "values": ["Low-Intermediate", "Medium", "High", ""],
        "chosen_value": "Low-Intermediate"
      },
      {
        "id": "5",
        "name": "Dynamic Drop down",
        "type": "drop_down2",
        "values": ["Low-Intermediate", "Medium", "High", ""],
        "chosen_value": "Low-Intermediate",
        "time": 1530164109066,
      },
      {"type": "header", "name": "Header", "id": "10029"},
      {
        "id": "1001",
        "name": "DVT",
        "type": "toggle",
        "values": ["On", "Off"],
        "chosen_value": null,
        "time": 1630164109066,
      },
      {"type": "header", "name": "Header2", "id": "10039"},
      {
        "id": "10056",
        "name": "ADVT",
        "type": "toggle",
        "values": ["On", "Off"],
        "chosen_value": "Off",
        "time": 1630164109056,
      },
      {
        "id": "1002",
        "name": "Static text",
        "type": "static_text",
        "chosen_value": "value which is long value",
        "description": "(description..)",
        "time": 1640164109066,
      },
      {
        "id": "10026",
        "name": "Long name of the line but a short",
        "type": "static_text",
        "chosen_value": "value",
        "description": "(description..)",
        "time": 1640164109066,
      },
      {
        "id": "1003",
        "name": "Blood Pressure with a long long long",
        "type": "edit_text",
        "chosen_value": "B",
        "time": 1640260609562,
        "description": "mmHg - with a long description",
      },
      {"type": "header", "name": "Header", "id": "10099"},
      {
        "id": "1004",
        "name": "Drop down",
        "type": "drop_down",
        "time": 1640264109066,
        "values": ["Low-Intermediate", "Medium", "High", ""],
        "chosen_value": "Low-Intermediate"
      },
  /*    {
        "id": "1005",
        "name": "Dynamic Drop down",
        "type": "drop_down2",
        "values": ["one", "two", "three"],
        "chosen_value": "one",
        "time": 1530164109066,
      },*/
      {"type": "header", "name": "Header", "id": "20029"},
      {
        "id": "2001",
        "name": "DVT",
        "type": "toggle",
        "values": ["On", "Off"],
        "chosen_value": null,
        "time": 1630164109066,
      },
      {"type": "header", "name": "Header2", "id": "20039"},
      {
        "id": "20056",
        "name": "ADVT",
        "type": "toggle",
        "values": ["On", "Off"],
        "chosen_value": "Off",
        "time": 1630164109056,
      },
      {
        "id": "2002",
        "name": "Static text",
        "type": "static_text",
        "chosen_value": "value which is long value",
        "description": "(description..)",
        "time": 1640164109066,
      },
      {
        "id": "20026",
        "name": "Long name of the line but a short",
        "type": "static_text",
        "chosen_value": "value",
        "description": "(description..)",
        "time": 1640164109066,
      },
      {
        "id": "2003",
        "name": "Blood Pressure with a long long long",
        "type": "edit_text",
        "chosen_value": "c",
        "time": 1640260609562,
        "description": "mmHg - with a long description",
      },
      {"type": "header", "name": "Header", "id": "20099"},
      {
        "id": "2004",
        "name": "Drop down",
        "type": "drop_down",
        "time": 1640264109066,
        "values": ["Low-Intermediate", "Medium", "High", ""],
        "chosen_value": "Low-Intermediate"
      },
/*      {
        "id": "2005",
        "name": "Dynamic Drop down",
        "type": "drop_down2",
        "values": ["one", "two", "three"],
        "chosen_value": "one",
        "time": 1530164109066,
      }*/
    ]
  };

  @override
  void initState() {
    super.initState();
    onValueChangeStream = _onUserController.stream.asBroadcastStream();
  }

  String buildDate(DateTime dateTime) {
    final now = DateTime.now();
    int diff = now.millisecondsSinceEpoch - dateTime.millisecondsSinceEpoch;

    if (diff < dayInMilliseconds) {
      return build24String(diff);
    } else if (diff >= dayInMilliseconds && diff < monthInMilliseconds) {
      return buildDaysString(diff);
    } else if (diff >= monthInMilliseconds && diff < yearInMilliseconds) {
      return buildMonthString(dateTime);
    }
    return dateTime.year.toString();
  }

  String buildDaysString(int diff) {
    int days = diff ~/ dayInMilliseconds;
    return "${days}d";
  }

  String buildMonthString(DateTime dateTime) {
    return DateFormat.MMMd().format(dateTime);
  }

  int yearInMilliseconds = 31556952000;
  int monthInMilliseconds = 2629800000;
  int weekInMilliseconds = 604800000;
  int dayInMilliseconds = 86400000;
  int hourInMilliseconds = 3600000;
  int minuteInMilliseconds = 60000;

  String build24String(int diff) {
    int hours = diff ~/ hourInMilliseconds;
    int minutes = (diff - (hours * hourInMilliseconds)) ~/ minuteInMilliseconds;
    return "${hours}h ${minutes}m";
  }

  Widget dateBuilder(int date, String id) {
    DateTime dateTime = DateTime.fromMillisecondsSinceEpoch(date);
    return Text(buildDate(dateTime));
  }
  Widget saveBarBuilder({required Function onSave, required Function onClose}) {
    return Padding(
      padding: const EdgeInsets.all(8.0),
      child: Row(
          children: [
            ElevatedButton(
                onPressed: ()=> onClose(),
                child: Text("close")),
            Spacer(),
            ElevatedButton(
                onPressed: ()=> onSave(),
                child: Text("save")),
          ]),
    );
  }

  List<String> list = ["Medium", "High"];
  List<String> toggleList = ["On", "Off"];

  int counter = 0;
  int toggle = 1;

  bool changeToGoodBad = true;

  Future<dynamic> _refresh() {
    //"chosen_value": "Low-Intermediate"
    if (changeToGoodBad) {
      toggleList = ['Good', "Bad"];
      json['widgets'][1]['values'] = toggleList;
      json['widgets'][1]['chosen_value'] = "Bad";
      json['widgets'][6]['chosen_value'] = "goodReally";
      json['widgets'][8]['chosen_value'] = "Medium";
      json['widgets'][9]['chosen_value'] = "Medium";

    } else {
      toggleList = ["On", "Off"];
      json['widgets'][1]['chosen_value'] = "On";
      json['widgets'][1]['values'] = toggleList;
      json['widgets'][6]['chosen_value'] = "bad";
      json['widgets'][8]['chosen_value'] = "High";
      json['widgets'][9]['chosen_value'] = "High";
    }

    changeToGoodBad = !changeToGoodBad;
    setState(() {});
    return Future.delayed(const Duration(seconds: 0));
  }

  @override
  Widget build(BuildContext context) {
    return Sizer(
      builder: (BuildContext context, Orientation orientation, DeviceType deviceType) {
        return Scaffold(
          appBar: AppBar(
            title: const Text('Floating Action Button'),
          ),
          body: RefreshIndicator(
            onRefresh: _refresh,
            child: JsonFormWithThemeBuilder(
                jsonWidgets: json)
                .setDateBuilderMethod(dateBuilder)//
             .registerComponent(DropDownParserCreator2())
                // .setDynamicFactory(MyWidgetParserFactory())
                .setStreamUpdates(onValueChangeStream)
                .setOnValueChanged((String d, dynamic s) async {
                  await Future.delayed(const Duration(seconds: 1));
                  return Future.value(true);
                })
               .setSaveBarBuilderMethod(saveBarBuilder)
               .setTheme(const DefaultTheme()).build(),
          ),
          floatingActionButton: FloatingActionButton(
            onPressed: () {
              counter++;
              if (counter % 5 == 1) {
                toggle++;
                _onUserController.add({}..["1"] = toggleList[toggle % 2]); // toggle
              }
              if (counter % 5 == 2) {
                _onUserController.add({}..["2"] = "updated" + Random().nextInt(10).toString()); // toggle
              }
              if (counter % 5 == 3) {
                _onUserController.add({}..["3"] = "Val" + Random().nextInt(100000).toString()); // toggle
              }
              if (counter % 5 == 4) {
                _onUserController.add({}..["4"] = list[toggle % 2]); // toggle
              }
              if (counter % 5 == 0) {
                _onUserController.add({}..["5"] = list[toggle % 2]); // toggle
              }
            },
            child: const Icon(Icons.navigation),
            backgroundColor: Colors.green,
          ),
        );
      },
    );
  }
}
